### PR TITLE
issue_30_logging - log exposed service names

### PR DESCRIPTION
**Pull request by robinroos**
If there is no match for the given service criteria then log the set of exposed service names.

### CODE REVIEW

File: src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java [hunk 1]
```diff
@@ -397,6 +397,21 @@ private boolean isSelf(NetworkAttachment networkAttachment) throws SocketExcepti
...
+			List<Service> allServices = docker.listServices();
+
+			StringBuilder sb = new StringBuilder();
+			String delim = "";
+			for (Service s : allServices) {
+				sb.append(delim).append(s.spec().name());
+				delim = ",";
+			}
+
+			logger.info("No service match for given criteria");
```
Comment (bitsofinfo) [inline src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java]: You don't need the verbosity of 2 log statements here, just combine into one one

This entire block needs to be configured by a new configuration option. OFF by default

**Comment by bitsofinfo** [comment]
This kind of debugging needs a new configuration option to enable it + documentation. Must be OFF by default. For those who have hundreds of services on a swarm, it can't be flooding this into people's logs without desiring it.

File: src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java (hunk 1 above)
Comment (robinroos) [reply src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java]: I was planning to drop the 2nd to DEBUG, hence having two.  DEBUG is off by default.

Comment (robinroos) [reply src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java]: Also, withdrawing this PR in order to resubmit with the TODO item addressed.

**Comment by bitsofinfo** [comment]
Please stop opening and closing PRs, you can just issue more commits in the already existing one

**Comment by robinroos** [comment]
ok, will use original branch and original PR.

**Comment by robinroos** [comment]
PR Build is Green.

The INFO message "No service match for given criteria" is correctly absent in Build #160 as the tests do not hit this specific issue for the NullPointerException.

If you can see fit to approve, please publish as RC13.

**Comment by bitsofinfo** [comment]
Please see: https://github.com/bitsofinfo/hazelcast-docker-swarm-discovery-spi/pull/34#issuecomment-468053557

**Comment by robinroos** [comment]
I noted your comment

> This kind of debugging needs a new configuration option to enable it + documentation. Must be OFF by default. For those who have hundreds of services on a swarm, it can't be flooding this into people's logs without desiring it.

That's why the list of exposed services is logged FINE.  And the very next thing any user is going to see is a NullPointerException.  We're close here, but I do not see the value of adding configuration options and associated documentation at this particular point.  I appreciate you are the product owner and your perspective is thus materially different to mine.

**Comment by robinroos** [comment]
At 4am UK time I will have to re-deploy the application stack to the swarm without cluster node discovery and with all services scaled to 1, until such time as I have resolved the issues we are facing.  This PR would give me the necessary info before I lose the currently-deployed "failing test case".

**Comment by bitsofinfo** [comment]
This was merged but refactored.

+ gated by new property to control the full service dump to logs regardless of severity
+ provided impl was moved as it would make unecessary repetitive calls to refetch all services, once for every configured name/label provided, rather than just 1x (after full evaluation)

**Comment by robinroos** [comment]
ok, thank you.

**Pull request merged**